mesonbuild/backend/ninjabackend.py

```diff
@@ -159,33 +159,41 @@ class Quoting(Enum):
     none = 3
 
 class NinjaCommandArg:
-    def __init__(self, s, quoting = Quoting.both):
+    def __init__(self, s: str, quoting: Quoting = Quoting.both) -> None:
         self.s = s
         self.quoting = quoting
 
-    def __str__(self):
+    def __str__(self) -> str:
         return self.s
 
     @staticmethod
-    def list(l, q):
+    def list(l: str, q: Quoting) -> T.List[NinjaCommandArg]:
         return [NinjaCommandArg(i, q) for i in l]
 
+@dataclass
 class NinjaComment:
-    def __init__(self, comment):
-        self.comment = comment
+    comment: str
 
-    def write(self, outfile):
+    def write(self, outfile: T.TextIO) -> None:
         for l in self.comment.split('\n'):
             outfile.write('# ')
             outfile.write(l)
             outfile.write('\n')
         outfile.write('\n')
 
+@dataclass
 class NinjaRule:
-    def __init__(self, rule, command, args, description,
-                 rspable = False, deps = None, depfile = None, extra = None,
-                 rspfile_quote_style: RSPFileSyntax = RSPFileSyntax.GCC):
-
+    name: str
+    command: T.List[T.Union[NinjaCommandArg, str]]
+    args: T.List[T.Union[NinjaCommandArg, str]]
+    description: str
+    rspable: bool = False # if a rspfile can be used
+    deps: T.Optional[str] = None # dep style 'gcc' or 'msvc'
```

Comment:
An enum is really meant for places where you want fast, typed comparison. This isn't used in any way other than as a direct string, so enums are the wrong approach.

The correct approach is probably to use a typing.Literal; I actually tried that but it turns out in many places we create a NinjaRule with deps parameter fed in by an if/elsed variable and Literal can't track that. Possibly it would work if I slapped in a bunch of annotations at various call sites, but there you have my premise...